### ISSUE TITLE
More progress required

### ISSUE THREAD

**Issue by Wapiti08** [shoaib-intro/logs_Analysis#1]
Can you please transform the raw data into dataframe first? I did not have much time to start from the beginning.

**Comment by shoaib-intro** [comment]
You mean data frame by `log_value_vector` file where we have following columns:
```
log message
log key
parameter value vector
```

**Comment by Wapiti08** [comment]
Yes. I have to say it is not able to finish the online learning part in your resp. Cuz we did not have the engine to parse logs. The data I process is from logstash with json format.

**Comment by shoaib-intro** [comment]
Okay, Let me do this!

**Comment by shoaib-intro** [comment]
1. Could you please check I have parsed structured files in `csv` and `pkl` format. 
2. I also converted parsed structured logs data into given image structure in `log_value_vector` file which is used by you in demo files.
Is this required by you?

![man](https://user-images.githubusercontent.com/56737996/164976509-06c44420-1c8c-4920-bdb7-fb189c2975bc.jpg)

**Comment by Wapiti08** [comment]
Yeah, you have to use pkl format in order to process the parameter value vector as list type. Now you can use labelencoder to prepcoessing the log key and train this column with Execution Path model. You can get all the detail from my resp

**Comment by Wapiti08** [comment]
Please train them with benign data.

**Comment by shoaib-intro** [comment]
Okay, just give me few minutes

**Comment by Wapiti08** [comment]
When you train for second model, remember padding the vectors to same length for one log key. Then train it with the right model (not in demo), there are some functions will calculate the rmse and print out the rmse value, the thresholds for false positive and true positive have been set. You will get both prediction values finally

**Comment by shoaib-intro** [comment]
While Training on Normal data I saw loss is 91% is this fine?


![show](https://user-images.githubusercontent.com/56737996/164977635-309ddb88-6c30-4da5-af62-4478496127b3.jpg)

**Comment by Wapiti08** [comment]
Yeah, if you have a concern about it, you can try to use normalization on training dataset. Then inverse_transfrom for the predicted result. It can help you decrease the loss

**Comment by shoaib-intro** [comment]
I have done with `Execution_Path sequences` could you please check it!

**Comment by Wapiti08** [comment]
Can you push your code as well, better with jupyter on colab. You can add the shared link, so I can see the cache and logic directly.

**Comment by shoaib-intro** [comment]
I have your code in scattered notebooks which I am combining to use....Initially i worked on your Demo folder and now with actual files which are updated in each folder as you mentioned.

**Comment by Wapiti08** [comment]
Can you push your code as well?

**Comment by shoaib-intro** [comment]
These are little bit messy :(

**Comment by Wapiti08** [comment]
No worries. I am not a teacher.

**Comment by shoaib-intro** [comment]
`pre-process` used for parsing into `log_value_vector` format
'`untitled`' used for both models
It would be hard for you to understand as its not cleaned notebook so please tell me I am here where you have queries!

**Comment by Wapiti08** [comment]
Hi,

I just check the code. You misunderstood the anomaly match function. It is used for parameter value vector model only. The execution model is quite straightforward. You just need to check the predicted results with the true result. No false positive

**Comment by shoaib-intro** [comment]
Hi Rahul, 
Thanks for you review!
Could you please tell me little bit more, we check first log key in your demo folder if its new its anomaly in second step we come across by checking parameters based anomaly if new parameter is there we consider it as anomaly...am I right? 
regarding logkey path execution in paper it predicts next log key sequence if not in k probability distributed predicted keys then its anomaly. 
if you dont mind

**Comment by Wapiti08** [comment]
The anomaly prediction order is like Exe ---> Para. The arrow starts only when the first prediction is good, which means if the prediction is a anomaly in Exe, it is the anomaly value.

**Comment by Wapiti08** [comment]
You do not need to think about this first, you just need to make sure your first model is right. Then check the second model. Finally, you can combine them together.

**Comment by shoaib-intro** [comment]
Hi Rahul,
I have shared notebook for parameters Anomaly detection under `test` folder.
I trained model using execution path after getting anomalies facing some errors in parameter anomaly file. could please have a look!

**Comment by shoaib-intro** [comment]
Hi @Wapiti08,

I have uploaded data and both models in notebook form while second model also from Demo folder. could you please check! I am available!

Best Regards

**Comment by Wapiti08** [comment]
Hi,

I have a basic check, I think you are right on the way.

**Comment by Wapiti08** [comment]
Do you have any further questions? My only recommendation is to simplify some processes.

**Comment by shoaib-intro** [comment]
> Hi @Wapiti08,
> 
> I have uploaded data and both models in notebook form while second model also from Demo folder. could you please check! I am available!
> 
> Best Regards

yes just please let check!

**Comment by shoaib-intro** [comment]
Actually I am done with first model but for second model I need encoded form of each parameters before  feeding to the model. 
were to find this encoding for each vector? I did encoding into npy folder by use of matrix_generation am I right in this case? 
Secondly I need the output file as you have `trace_df` in parameter folder how to generate that file? as in `parameter_Value_Vector` you have mentioned this for only one logkey as shwon below but I want for all log  key!

![image](https://user-images.githubusercontent.com/56737996/167253166-bb4efce1-cc0d-44f8-ba36-fd36b86a5a9f.png)

**Comment by Wapiti08** [comment]
Can you list your questions with - ?

**Comment by Wapiti08** [comment]
I do not entirely understand your problems.

**Comment by shoaib-intro** [comment]
> I do not entirely understand your problems.

yes please I am writing in details!

**Comment by shoaib-intro** [comment]
1. One we detect anomalies by Execution path of log keys then we have to `check each` logkey parameter which are detected as anomalies from first model right?  lie below in `path_pred` columns we have two logkeys detected as anomalies:

![image](https://user-images.githubusercontent.com/56737996/167253436-2157d7d8-54e7-42f5-9d5c-fa76af4f7247.png)

2. Then in second step we will check parameters anomalies for each `logkey` which are detected anomalies(=means marked 1) during first model. (like here we have to check parameter anomalies for two logkeys) right?
For detecting parameters anomalies when I run you second model(`Parameter_value_vector`) it works on only for one logkey to detect parameters anomalies  **how to detect parameters anomalies for all those logkeys which are marked as 1 in path_pred columns?** 
This code detect parameter anomalies for only below mentioned one logkey (as shown below) but I need to detect for many logkeys parameters anomalies. where you did this step.
![image](https://user-images.githubusercontent.com/56737996/167253712-d7a77bee-66fe-436a-a33d-3fb981844d35.png)

I want output file like you have below (where `para_pred` column checks for every logkey parameters):
![image](https://user-images.githubusercontent.com/56737996/167253856-fcb0bd81-91ff-481e-89e7-8fb0ca976838.png)

> So I only have one question taking input `structured logs` parsed by parser and detect parameters anomalies for each logkey detected as anomalies by first model to get `trace_df` file as output final marked anomalies (your `Parameter_Value_Vector` does this but for only one logkey which I give manually)?

Please tell me if its understandable otherwise I will elaborate it!

**Comment by Wapiti08** [comment]
- For your first question, if the detection from first model is an anomaly, which means there is no need to go for the second model. I think you misunderstood the point. The detection goes for the second model if the first detection is benign. 

- For your second question: After you found the prediction first model is benign, you have to find the corresponding log key and original index of that row (one parameter vector). 

From my personal implementation, I predict the dataframe with two models in the same time. I got result like:

path_pred, para_pred, final pred
0             ,  1             , 1
1             , 0              , 1
1             ,  1             , 1
 0            , 0              , 0
 
 By doing this, there is no need to check the results from second model based on the first model.

**Comment by shoaib-intro** [comment]
Alright, I got your point when to go for second model, thanks making it clear!
Can I ask script for second model which checks all benign from first and detects using second model and predicts output like you've done on your personal implementation as shown below:

path_pred, **para_pred**, final pred
0 , **1**, 1
1 , **0** , 1
1 , **1** , 1
0 , **0** , 0

**Comment by Wapiti08** [comment]
For privacy reasons (company project), I cannot give that to you.

I can only tell the core point is to make use of the original indexes and orders in a key group. (Of course, you need to filter the first few rows and also bypass the processing for the key group without enough length)

**Comment by shoaib-intro** [comment]
How to increase model performance:
> Related to first model
- I trained **first** model on `125k` normal loglines and tested on **same** loglines but model detected `767` anomalies with when I tested with `probability 0.1` when tested for same data with `probability 0.0001` it detected 123 anomalies (shown below) so what is **effect of probability on anomalies detection** (because detected anomalies are not actual anomalies as  trained model on same data and tested same but anomalies are coming)? 

> Related to online learning
- When I trained online this model above model with `probability 0.5` it has no effect on model learning but when I train online with `probability = 0.0005` then model reduces anomalies from 767 to 323 so, again what is probability **effect on online training**?  


What do you think what is solution to reduce these FP? 

> Related to Deep log paper
- In `Deeplog` paper it is mentioned that model learns on 1% data of logs data but in actual case is not like this what do you think about this?

**Comment by Wapiti08** [comment]
- Firstly, you have to introduce human check on what is really an anomaly or not. Then take this into online learning.
- The probability is not a core point in this project. 
- Answered your question.

**Comment by shoaib-intro** [comment]
Its mean only anomalies should be considered for online learning right. 
I like you way professional coding I am not that level of expert but I will ask you few recommendation and learning material to grab this level of expertise, please suggest (online courses, articles or any training always welcome to get handy with this level of python expertise and modelling)!!!!  
Could you please share your social link or whatsapp any!

**Comment by shoaib-intro** [comment]
have you done **workflow** part?

**Comment by Wapiti08** [comment]
> 
I am far away from professional coding. I would recommend the [course](https://www.udemy.com/course/deployment-of-machine-learning-models/) for machine learning skill. The way to the expert or professional level is a gradual process. Read the official documents as well. 

I did not implement that part.

**Comment by shoaib-intro** [comment]
please share your linkedin I tried with your email didn't find it

**Comment by Wapiti08** [comment]
We have connected [linkedin](https://www.linkedin.com/in/newt-tan-b8ba7a155/)
